zkevm-circuits/src/pi_circuit.rs

```diff
@@ -0,0 +1,978 @@
+        }
+        tx_vals
+    }
+
+    /// Returns struct with the extra values
+    pub fn get_extra_values(&self) -> ExtraValues {
+        ExtraValues {
+            // block_hash: self.extra.eth_block.hash.unwrap_or_else(H256::zero),
+            state_root: self.extra.eth_block.state_root,
+            parent_block_hash: self.extra.eth_block.parent_hash,
+        }
+    }
+}
+
+/// Config for PiCircuit
+#[derive(Clone, Debug)]
+pub struct PiCircuitConfig<F: Field, const MAX_TXS: usize, const MAX_CALLDATA: usize> {
+    q_block_table: Selector,
+    block_value: Column<Advice>,
+
+    q_tx_table: Selector,
+    tx_id: Column<Advice>,
+    tag: Column<Fixed>,
+    index: Column<Advice>,
+    tx_value: Column<Advice>,
+
+    raw_public_inputs: Column<Advice>,
+    rpi_rlc_acc: Column<Advice>,
+    rand_rpi: Column<Advice>,
+    q_not_end: Selector,
+    q_end: Selector,
+
+    pi: Column<Instance>, // p, rand_pi, chain_ID, state_root, prev_state_root
+    _marker: PhantomData<F>,
+}
+
+impl<F: Field, const MAX_TXS: usize, const MAX_CALLDATA: usize>
+    PiCircuitConfig<F, MAX_TXS, MAX_CALLDATA>
+{
+    fn new(meta: &mut ConstraintSystem<F>) -> Self {
+        let q_block_table = meta.selector();
+        // BlockTable
+        let block_value = meta.advice_column();
+
+        let q_tx_table = meta.selector();
+        // Tx Table
+        let tx_id = meta.advice_column();
+        let tag = meta.fixed_column();
+        let index = meta.advice_column();
+        let tx_value = meta.advice_column();
+
+        let raw_public_inputs = meta.advice_column();
+        let rpi_rlc_acc = meta.advice_column();
+        let rand_rpi = meta.advice_column();
+        let q_not_end = meta.selector();
+        let q_end = meta.selector();
+
+        let pi = meta.instance_column();
+
+        meta.enable_equality(raw_public_inputs);
+        meta.enable_equality(rpi_rlc_acc);
+        meta.enable_equality(rand_rpi);
+        meta.enable_equality(pi);
+
+        // 0.0 rpi_rlc_acc[0] == RLC(raw_public_inputs, rand_rpi)
+        meta.create_gate(
+            "rpi_rlc_acc[i] = rand_rpi * rpi_rlc_acc[i+1] + raw_public_inputs[i] ",
+            |meta| {
+                // row.rpi_rlc_acc == q_not_end * row_next.rpi_rlc_acc * row.rand_rpi +
+                // row.raw_public_inputs
+                let q_not_end = meta.query_selector(q_not_end);
+                let cur_rpi_rlc_acc = meta.query_advice(rpi_rlc_acc, Rotation::cur());
+                let next_rpi_rlc_acc = meta.query_advice(rpi_rlc_acc, Rotation::next());
+                let rand_rpi = meta.query_advice(rand_rpi, Rotation::cur());
+                let raw_public_inputs = meta.query_advice(raw_public_inputs, Rotation::cur());
+
+                vec![
+                    q_not_end * (next_rpi_rlc_acc * rand_rpi + raw_public_inputs - cur_rpi_rlc_acc),
+                ]
+            },
+        );
+        meta.create_gate("rpi_rlc_acc[last] = raw_public_inputs[last]", |meta| {
```

Comment:
It wasn't missing, but I needed to adapt the constraint as specified in the spec and add this constraint.
The constraint in the spec is the following:
```
        row.rpi_rlc_acc == q_not_end * row_next.rpi_rlc_acc * row.rand_rpi + row.raw_public_inputs
```
which for the last row, having `q_not_end  = 0` ensures: `row.rpi_rlc_acc == row.raw_public_inputs`.
However, the problem  I run into with this approach is that the constraint fails for the last row because it tries to access the value in the next rotation, even though it is multiplied by 0. This is why I split the selectors into `q_end` and `q_not_end`, so that this constraint was deactivated in the last row and I thus avoided the rotation issue.
I'd like to find a more elegant workaround for this issue so that we could eliminate one of these selectors.

Reply:
After talking with @han0110 , we decided that is likely an ok solution, since the selectors `q_end` and `q_not_en` will be optimized internally.